File: server/controllers/jurisdictionController.js
const jurisdictionModel = require("../models/jurisdictionModel");
const pool = require("../config/db");

// Controller: Fetch all States
exports.getStates = async (req, res) => {
  try {
    const states = await jurisdictionModel.getStates();
    res.status(200).json({
      success: true,
      message: "States fetched successfully",
      data: states,
    });
  } catch (error) {
    console.error("Error fetching states:", error);
    res.status(500).json({
      success: false,
      message: "Failed to fetch states",
      error: error.message,
    });
  }
};

// Controller: Fetch Divisions by State
exports.getDivisionsByState = async (req, res) => {
  try {
    const { stateId } = req.params;
    if (!stateId)
      return res.status(400).json({ success: false, message: "stateId is required" });

    const divisions = await jurisdictionModel.getDivisionsByState(stateId);
    res.status(200).json({
      success: true,
      message: "Divisions fetched successfully",
      data: divisions,
    });
  } catch (error) {
    console.error("Error fetching divisions:", error);
    res.status(500).json({
      success: false,
      message: "Failed to fetch divisions",
      error: error.message,
    });
  }
};

// Controller: Fetch Districts by Division
exports.getDistrictsByDivision = async (req, res) => {
  try {
    const { divisionId } = req.params;
    if (!divisionId)
      return res.status(400).json({ success: false, message: "divisionId is required" });

    const districts = await jurisdictionModel.getDistrictsByDivision(divisionId);
    res.status(200).json({
      success: true,
      message: "Districts fetched successfully",
      data: districts,
    });
  } catch (error) {
    console.error("Error fetching districts:", error);
    res.status(500).json({
      success: false,
      message: "Failed to fetch districts",
      error: error.message,
    });
  }
};

// Controller: Fetch Blocks by District
exports.getBlocksByDistrict = async (req, res) => {
  try {
    const { districtId } = req.params;
    if (!districtId)
      return res.status(400).json({ success: false, message: "districtId is required" });

    const blocks = await jurisdictionModel.getBlocksByDistrict(districtId);
    res.status(200).json({
      success: true,
      message: "Blocks fetched successfully",
      data: blocks,
    });
  } catch (error) {
    console.error("Error fetching blocks:", error);
    res.status(500).json({
      success: false,
      message: "Failed to fetch blocks",
      error: error.message,
    });
  }
};

// Controller: Fetch Clusters by Block
exports.getClustersByBlock = async (req, res) => {
  try {
    const { blockId } = req.params;
    if (!blockId)
      return res.status(400).json({ success: false, message: "blockId is required" });

    const clusters = await jurisdictionModel.getClustersByBlock(blockId);
    res.status(200).json({
      success: true,
      message: "Clusters fetched successfully",
      data: clusters,
    });
  } catch (error) {
    console.error("Error fetching clusters:", error);
    res.status(500).json({
      success: false,
      message: "Failed to fetch clusters",
      error: error.message,
    });
  }
};

// Controller: Fetch Institutes by Cluster
exports.getInstitutesByCluster = async (req, res) => {
  try {
    const { clusterId } = req.params;
    if (!clusterId)
      return res.status(400).json({ success: false, message: "clusterId is required" });

    const institutes = await jurisdictionModel.getInstitutesByCluster(clusterId);
    res.status(200).json({
      success: true,
      message: "Institutes fetched successfully",
      data: institutes,
    });
  } catch (error) {
    console.error("Error fetching institutes:", error);
    res.status(500).json({
      success: false,
      message: "Failed to fetch institutes",
      error: error.message,
    });
  }
};

exports.getJurisName = async (req, res) => {
  try {
    const { juris_code } = req.params;
    if (!juris_code) {
      return res.status(400).json({ success: false, message: "juris_code is required" });
    }
    const jurisName = await jurisdictionModel.getJurisName(juris_code);
    res.status(200).json({
      success: true,
      message: "Jurisdiction name fetched successfully",
      data: jurisName,
    });
  } catch (error) {
    console.error("Error fetching jurisdiction name:", error);
    res.status(500).json({
      success: false,
      message: "Failed to fetch jurisdiction name",
      error: error.message,
    });
  }
};

File: java-backend/imas-backend/src/main/java/com/rcf/imas/modules/masterdata/web/JurisdictionController.java
package com.rcf.imas.modules.masterdata.web;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.rcf.imas.modules.masterdata.persistence.JurisdictionRepository;
import com.rcf.imas.platform.error.ApiException;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api")
// Node's jurisdiction/districts/institutes/juris-names mounts had NO auth (public). These are read-only
// geography/reference lookups consumed by admin screens (via useJurisData) today; gated isAuthenticated()
// -- role-agnostic so no authenticated caller is ever 403'd -- rather than hasRole('ADMIN'), since locking
// low-sensitivity reference data to ADMIN risks breaking any shared/non-admin consumer. Stricter than Node.
@PreAuthorize("isAuthenticated()")
class JurisdictionController {

    record JurisNamesRequest(@JsonProperty("districtIds") List<Long> districtIds,
                             @JsonProperty("blockIds") List<Long> blockIds,
                             @JsonProperty("instituteIds") List<String> instituteIds) {}

    private final JurisdictionRepository repo;

    JurisdictionController(JurisdictionRepository repo) { this.repo = repo; }

    @GetMapping("/states")
    public List<Map<String, Object>> states() { return repo.states(); }

    @GetMapping("/divisions-by-state/{stateId}")
    public List<Map<String, Object>> divisions(@PathVariable String stateId) {
        return repo.childrenOf("DIVISION", stateId);
    }

    @GetMapping("/districts-by-division/{divisionId}")
    public List<Map<String, Object>> districts(@PathVariable String divisionId) {
        return repo.childrenOf("EDUCATION DISTRICT", divisionId);
    }

    @GetMapping("/blocks-by-district/{districtId}")
    public List<Map<String, Object>> blocks(@PathVariable String districtId) {
        return repo.childrenOf("BLOCK", districtId);
    }

    @GetMapping("/clusters-by-block/{blockId}")
    public List<Map<String, Object>> clusters(@PathVariable String blockId) {
        return repo.childrenOf("CLUSTER", blockId);
    }

    @GetMapping("/institutes-by-cluster/{clusterId}")
    public List<Map<String, Object>> institutesByCluster(@PathVariable String clusterId) {
        return repo.institutesByCluster(clusterId);
    }

    @GetMapping("/juris-name/{jurisCode}")
    public Map<String, Object> jurisName(@PathVariable String jurisCode) {
        // Node returns rows[0]; missing code returned undefined -> empty body. Preserve leniently:
        return repo.jurisName(jurisCode).orElse(Map.of());
    }

    @GetMapping("/districts/all")
    public List<Map<String, Object>> allDistricts() { return repo.allDistricts(); }

    @GetMapping("/institutes/all")
    public List<Map<String, Object>> allInstitutes() { return repo.allInstituteNames(); }

    @GetMapping("/institutes/search")
    public List<Map<String, Object>> searchInstitutes(@RequestParam(required = false) String query) {
        if (query == null || query.isBlank()) {
            throw ApiException.error(400, "Missing query parameter");
        }
        return repo.searchInstitutes(query);
    }

    @PostMapping("/juris-names")
    public Map<String, Object> jurisNames(@RequestBody JurisNamesRequest req) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("districts", repo.namesByType("EDUCATION DISTRICT", req.districtIds()));
        body.put("blocks", repo.namesByType("BLOCK", req.blockIds()));
        body.put("institutes", repo.instituteNamesByDise(req.instituteIds()));
        return body;
    }
}
